filters/auth/grantcookie.go

```diff
@@ -0,0 +1,134 @@
+package auth
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+	"time"
+
+	"github.com/zalando/skipper/secrets"
+	"golang.org/x/oauth2"
+)
+
+type cookie struct {
+	AccessToken  string    `json:"access_token"`
+	RefreshToken string    `json:"refresh_token"`
+	Expiry       time.Time `json:"expiry,omitempty"`
+	RefreshAfter time.Time `json:"refresh_after,omitempty"`
+}
+
+func refreshAfter(expiry time.Time) time.Time {
+	now := time.Now()
+	if now.After(expiry) {
+		return now
+	}
+
+	d := expiry.Sub(now)
+	d /= 10
+	if d < time.Minute {
+		d = time.Minute
+	}
+
+	return now.Add(d)
+}
+
+func decodeCookie(cookieHeader string, config OAuthConfig) (c *cookie, err error) {
+	var eb []byte
+	if eb, err = base64.StdEncoding.DecodeString(cookieHeader); err != nil {
+		return
+	}
+
+	var encryption secrets.Encryption
+	if encryption, err = config.Secrets.GetEncrypter(secretsRefreshInternal, config.SecretFile); err != nil {
+		return
+	}
+
+	var b []byte
+	if b, err = encryption.Decrypt(eb); err != nil {
+		return
+	}
+
+	err = json.Unmarshal(b, &c)
+	return
+}
+
+func (c *cookie) isAccessTokenExpired() bool {
+	now := time.Now()
+	return now.After(c.Expiry)
+}
+
+// extractCookie removes and returns the OAuth Grant token cookie from a HTTP request.
+// The function supports multiple cookies with the same name and returns
+// the best match (the one that decodes properly).
+// The client may send multiple cookies if a parent domain has set a
+// cookie of the same name.
+// The grant token cookie is extracted so it does not get exposed to untrusted downstream
+// services.
+func extractCookie(request *http.Request, config OAuthConfig) (cookie *cookie, err error) {
+	old := request.Cookies()
+	new := make([]*http.Cookie, 0, len(old))
+
+	for i, c := range old {
+		if c.Name == config.TokenCookieName {
+			cookie, _ = decodeCookie(c.Value, config)
+			if cookie != nil {
+				new = append(new, old[i+1:]...)
+				break
+			}
+		}
+		new = append(new, c)
+	}
+
+	if cookie != nil {
+		request.Header.Del("Cookie")
+		for _, c := range new {
+			request.AddCookie(c)
+		}
+		return cookie, nil
+	}
+	return nil, http.ErrNoCookie
+}
+
+func CreateCookie(config OAuthConfig, host string, t *oauth2.Token) (*http.Cookie, error) {
+	c := cookie{
+		AccessToken:  t.AccessToken,
+		RefreshToken: t.RefreshToken,
+		Expiry:       t.Expiry,
+	}
+
+	if !config.DisableRefresh {
+		c.RefreshAfter = refreshAfter(t.Expiry)
```

Comment:
Do we need this and the DisableRefresh config?
I think if we have no RefreshToken then we do not refresh, if we have one we refresh.
Any reason to disable this on purpose?

Reply:
we can drop it. I don't recall the reason for this. @muaazsaleem by chance, do you remember if we added this on purpose in the preliminary prototype?

Reply:
I also don't see why we can't drop it.